programs/gamma/src/fees/dynamic_fee.rs

```diff
@@ -204,65 +197,58 @@ impl DynamicFee {
                 index: index as u16,
                 observation: *observation,
             })
             .collect::<Vec<_>>();
 
+        // Sort observations by timestamp (newest first)
         descending_order_observations.sort_by(|a, b| {
             { b.observation.block_timestamp }.cmp(&{ a.observation.block_timestamp })
         });
-        
 
+        // Need at least 2 observations to calculate prices
         if descending_order_observations.len() < 2 {
             // Not enough data points to compute TWAP
             return Ok((0, 0, 0));
         }
 
-        // For TWAP: use first and last observations
-        let first_obs = descending_order_observations.first().unwrap();
-        let last_obs = descending_order_observations.last().unwrap();
-        
-        let total_time_delta = last_obs
+        // For TWAP: use first and last observations within our window
+        // Get newest and oldest observations in our filtered set
+        let newest_obs = descending_order_observations.first().unwrap();
+        let oldest_obs = descending_order_observations.last().unwrap();
+
+        // Calculate time delta using the correct start time
+        let total_time_delta = newest_obs
             .observation
             .block_timestamp
-            .saturating_sub(first_obs.observation.block_timestamp) as u128;
+            .saturating_sub(oldest_obs.observation.block_timestamp)
+            as u128;
+
         if total_time_delta == 0 {
             return Ok((0, 0, 0));
         }
 
-        // Calculate TWAP directly from cumulative prices
-        let twap_price = last_obs
+        // Calculate TWAP using real observations only
+        let twap_price = newest_obs
             .observation
             .cumulative_token_0_price_x32
-            .checked_sub(first_obs.observation.cumulative_token_0_price_x32)
+            .checked_sub(oldest_obs.observation.cumulative_token_0_price_x32)
             .ok_or(GammaError::MathOverflow)?
             .checked_div(total_time_delta)
             .ok_or(GammaError::MathOverflow)?;
 
         // Iterate to find min/max spot prices
         for observation_with_index in descending_order_observations {
-            let is_in_observation_window = current_time
-                .saturating_sub(observation_with_index.observation.block_timestamp)
-                <= window;
-
-            if !is_in_observation_window {
-                // they are already in descending order of block timestamp.
-                break;
-            }
             let last_observation_index = if observation_with_index.index == 0 {
                 OBSERVATION_NUM - 1
             } else {
                 observation_with_index.index as usize - 1
             };
 
-            // if last observation is not valid, skip this observation
-            if observation_state.observations[last_observation_index].block_timestamp == 0 {
-                continue;
-            }
-
```

Comment:
This is still possible.

Lets keep this check.

This is possible for example:
We start from 0, we take 5 obervartation at time t1,t2,t3,t4,t5
[t1,t2,t3,t4,t5]

in this case the observation before t1 is uninitalized i.e with block_timestamp=0
As the index of observation is 0 and the last for that one is 99th observation.

Reply:
observation.block_timestamp != 0 condition is checked while getting the descending_order_observations vector